### PR TITLE
Ruby Tic Tac Toe: Miletsone 3

### PR DESCRIPTION
**Pull request by felix-vondee**
In this milestone, we implemented the logic to determine the winner of the game. The following were implemented:

1) Removed dummy data from the previous milestone
2) Included logic to update the board after each user plays
3) Included the logic to determine the winner of the game

Note: In our codebase, we did not notice any necessary opportunities to use attr readers, writers, or accessors. All of the game logic worked as expected, so we did not see the need to use the above-mentioned.

### CODE REVIEW

File: bin/main.rb [hunk 1]
```diff
@@ -1,41 +1,41 @@
 #!/usr/bin/env ruby
 
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
 class TicTacToe
+  WINNING_COMBINATIONS = [[0, 1, 2], [3, 4, 5], [6, 7, 8],
```
Comment (pacyL2K19) [inline bin/main.rb]: # Highlight
> Like the logic :+1:

**Review by pacyL2K19 — CHANGES_REQUESTED** [review]
# Status: Chages Requested :stop_sign: 

Hello @felix-vondee  and @zilton7 
You made great work so far :+1: the game logic works perfectly and I enjoyed playing the game :smile: 
But stills some changes you need to update before getting approved for this project
- [x] Please split the game logic in different files, making sure you follow the DRY rule
- [x] Make sure  you only put commands like `puts` and `gets.chomp` in the `main.rb` file 
- [x] Think about classes (like Player) to make your project more scalable, remembering to always put each class in a specific file for easy debugging

Work on these changes before you request another Code review request, feel free to contact me on Slack or LinkedIn [Pacifique Linjanja](https://www.linkedin.com/in/pacifique-linjanja/) or even schedule  meet on zoom in case you need more explanation [Here](https://calendly.com/d/gw4f-8wj9/minutes-with-tse)
Happy coding :man_technologist:

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @felix-vondee and @zilton7 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

Kindly work on the suggested changes  👇.
- [ ] You game is not working properly please update the win, lose and draw cases. you can see the on the winning case the Draw game message is shown
draw

- [ ] Make sure you only `put` commands like `puts` and `gets.chomp` should be in the main.rb file
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  def start
+    assign_players
+    game_loop
+  end
+
+  def display_turn
+    puts "it's #{@current_player.name}'s turn!"
+  end
```
Comment (rukundoeric) [inline lib/game.rb]: You should not use `puts` keyword here.  All `puts, gets or print` keywords should only be used in `main.rb`.

File: lib/game.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    actual_array = actual_array.each_slice(3).to_a
+    actual_array.each do |el|
+      return @current_player if el == %w[X X X] || el == %w[O O O]
+    end
+    nil
+  end
+
+  def draw
+    puts 'It\'s a DRAW'
+  end
```
Comment (rukundoeric) [inline lib/game.rb]: You should not use `puts` keyword here.  All `puts, gets or print` keywords should only be used in `main.rb`.

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  def game_over(winner)
+    puts 'GAME OVER!'
+    @board.display
+    if winner
+      puts "'#{winner.name}' is the Winner!"
```
Comment (rukundoeric) [inline lib/game.rb]: You should not use `puts` keyword here.  All `puts, gets or print` keywords should only be used in `main.rb`.

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve 
  - [x] [MANDATORY] All `puts, gets or print` keywords should only be used in `main.rb`. Kindly, refactor your code so that only `main.rb` contains those keywords.
  - [x] [MANDATORY] Kindly remove all commented codes.    

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

File: bin/main.rb [hunk 5]
```diff
@@ -1,95 +1,52 @@
...
   def initialize
-    @turns = 0
-    @board = [1, 2, 3, 4, 5, 6, 7, 8, 9]
-    assign_players
-    set_current_player
-    display_board
+    player_a = Player.new(input_names('X'))
+    player_b = Player.new(input_names('O'))
+    @game = Game.new(player_a, player_b)
+    # @game.assign_players
```
Comment (rukundoeric) [inline bin/main.rb:11]: this commented code should be removed

**Review by rukundoeric — APPROVED** [review]
# Status: 🍾 PROJECT APPROVED 🍾
_Excellent job on this project guys. 🎆 Keep it up._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve
  -  [x] [MANDATORY] make sure you remove all commented codes before you request review for the next milestone.

_Congratulations on the splendid job **@team** 👍
You can now merge this PR and submit the project completion form 🚀 . I wish you all the very best in your next tasks 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/),
Slack: Rukundo Eric**

**Pull request merged**